### ISSUE TITLE
buildEndpointRequestPayload EXC_BAD_ACCESS_KERN_INVALID_ADDRESS

### ISSUE THREAD

**Issue by daniel-beard** [aws-amplify/aws-sdk-ios#4551]
**Describe the bug**
Bad access crash that happens infrequently, to a small percentage of users.

**To Reproduce**
No reproduction steps known yet, this crash occurs at a (relatively) low rate with users.

**Observed Behavior**

**Expected Behavior**
Should not crash

**Stack Trace**

EXC_BAD_ACCESS_KERN_INVALID_ADDRESS

```
Crashed: com.amazonaws.AWSPinpointEventRecorder
0  libobjc.A.dylib                0x3680 objc_release + 16
1  libobjc.A.dylib                0x3680 objc_release_x0 + 16
2  MyCoolApp                      0x74e490 -[AWSPinpointEventRecorder buildEndpointRequestPayload:] + 1160 (AWSPinpointEventRecorder.m:1160)
3  MyCoolApp                      0x74efa8 -[AWSPinpointEventRecorder putEventsRequestForEvents:endpointProfile:] + 1268 (AWSPinpointEventRecorder.m:1268)
4  MyCoolApp                      0x74c90c -[AWSPinpointEventRecorder putEvents:error:endpointProfile:] + 1035 (AWSPinpointEventRecorder.m:1035)
5  MyCoolApp                      0x74c050 -[AWSPinpointEventRecorder putEvents:error:] + 963 (AWSPinpointEventRecorder.m:963)
6  MyCoolApp                      0x749f40 __46-[AWSPinpointEventRecorder submitBatchEvents:]_block_invoke + 744 (AWSPinpointEventRecorder.m:744)
7  MyCoolApp                      0x72a420 __63-[AWSTask continueWithExecutor:successBlock:cancellationToken:]_block_invoke + 398 (AWSTask.m:398)
8  MyCoolApp                      0x729fe4 __56-[AWSTask continueWithExecutor:block:cancellationToken:]_block_invoke + 331 (AWSTask.m:331)
9  libdispatch.dylib              0x63850 _dispatch_call_block_and_release + 24
10 libdispatch.dylib              0x647c8 _dispatch_client_callout + 16
11 libdispatch.dylib              0x3f854 _dispatch_lane_serial_drain$VARIANT$armv81 + 604
12 libdispatch.dylib              0x402e4 _dispatch_lane_invoke$VARIANT$armv81 + 380
13 libdispatch.dylib              0x4a000 _dispatch_workloop_worker_thread + 612
14 libsystem_pthread.dylib        0x1b50 _pthread_wqthread + 284
15 libsystem_pthread.dylib        0x167c start_wqthread + 8
```

**Code Snippet**
- NOTE: I am *only* using pinpoint. Here's the setup code I'm using:

```swift
    let credentialsProvider = AWSCognitoCredentialsProvider(regionType: PinpointEndpoint.REGION,
                                                                identityPoolId: "{COGNITO_KEY_REDACTED}")
    let configuration = AWSPinpointConfiguration(appId: "{APP_ID_REDACTED}",
                                                    launchOptions: nil)
    configuration.serviceConfiguration = AWSServiceConfiguration(region: PinpointEndpoint.REGION,
                                                                    credentialsProvider: credentialsProvider)
    configuration.targetingServiceConfiguration = configuration.serviceConfiguration
    self.pinpoint = AWSPinpoint(configuration: configuration)

    let endpointProfile = pinpoint.targetingClient.currentEndpointProfile()
    if endpointProfile.address == nil {
        endpointProfile.address = UUID().uuidString
        pinpoint.targetingClient.update(endpointProfile)
    }

    pinpoint.analyticsClient.submitEvents()
    self.submissionTimer = SubmissionTimer(pinpoint: pinpoint, notificationCenter: NotificationCenter.default)
```

**Areas of the SDK you are using (AWSMobileClient, Cognito, Pinpoint, IoT, etc)?**
- Pinpoint

**Environment(please complete the following information):**
 - SDK Version: 2.28.5
 - Dependency Manager: Cocoapods
 - Swift Version : 5.7
 - Xcode Version: 14.2

**Device Information (please complete the following information):**
 - Device: Real devices
 - iOS Version: iOS 14 -> iOS 16
 - Specific to simulators: Not observed on simulator - because I don't have reproduction steps yet.

**Comment by daniel-beard** [comment]
After more testing, this appears to be a deadlock caused because of how `AWSPinpointEventRecorder.submitAllEvents` is implemented.

Although `submitAllEvents` has an `@synchronized(self.lock)` call within it, this method is NOT THREAD SAFE, and prone to deadlocking applications.

**Comment by ruisebas** [comment]
Hi @daniel-beard ! Thanks for opening this issue and for the investigation.

We will take a look into addressing this problem and post any updates in here.

**Comment by ruisebas** [comment]
@daniel-beard This issue has been fixed in [2.30.2](https://github.com/aws-amplify/aws-sdk-ios/releases/tag/2.30.2).

Please upgrade your Pinpoint dependency and let us know if you still have instances of this crash.

Thanks!
